src/validation.cpp

```diff
@@ -2446,6 +2479,14 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         const CBlockIndex *pindexFork;
         ConnectTrace connectTrace;
         bool fInitialDownload;
+        { // TODO: Tempoarily ensure that mempool removals are notified before
+          // connected transactions.  This shouldn't matter, but the abandoned
+          // state of transactions in our wallet is currently cleared when we
+          // receive another notification and there is a race condition where
+          // notification of a connected conflict might cause an outside process
+          // to abandon a transaction and then have it inadvertantly cleared by
+          // the notification that the conflicted transaction was evicted.
+        MemPoolConflictRemovalTracker mrt(mempool);
```

Comment:
I believe you need some lock held here so that no two threads could be calling mempool.NotifyEntryRemoved.connect at the same time, however I agree you dont want to SyncTransaction with cs_main held. You could add a static mutex here, or you could create the object with cs_main, then std::move it to a dummy which will be destroyed without cs_main.